### PR TITLE
Log more exceptions and ensure looping, change telemetry timing metric

### PR DESCRIPTION
**Pull request by drwill-ms**

### CODE REVIEW

**Review by patilsnr — APPROVED** [review]

File: e2e/LongHaul/service/IotHub.cs [hunk 1]
```diff
@@ -171,25 +182,42 @@ public async Task ReceiveFileUploadNotificationsAsync(CancellationToken ct)
...
+                await s_blobContainerClient.DeleteBlobIfExistsAsync(fileUploadNotification.BlobName).ConfigureAwait(false);
+                return AcknowledgementType.Complete;
+            }
+
+            async Task FileUploadNotificationErrors(ErrorContext errorContext)
+            {
+                Exception exToLog = errorContext.IotHubServiceException == null
+                    ? errorContext.IOException
+                    : errorContext.IotHubServiceException;
+                _logger.Trace($"Error processing C2D message.\n{exToLog}");
```
Comment (brycewang-microsoft) [inline e2e/LongHaul/service/IotHub.cs:212]:
```suggestion
                _logger.Trace($"Error processing file uploading notifications.\n{exToLog}");
```

**Review by brycewang-microsoft — APPROVED** [review]

File: iothub/service/src/Exceptions/IotHubServiceException.cs [hunk 2]
```diff
@@ -91,6 +91,18 @@ protected IotHubServiceException(SerializationInfo info, StreamingContext contex
         /// <seealso href="https://docs.microsoft.com/rest/api/iothub/common-error-codes"/>.
         public IotHubServiceErrorCode ErrorCode { get; }
 
+        /// <inheritdoc/>
+        public override string ToString()
+        {
+            var sb = new StringBuilder($"Message: {Message}\nErrorCode: {ErrorCode}, IsTransient: {IsTransient}");
```
Comment (brycewang-microsoft) [inline iothub/service/src/Exceptions/IotHubServiceException.cs:97]: As we have `StatusCode` and `ErrorCode` properties in "IotHubServiceException", I would like to include both of them together. 
```suggestion
            var sb = new StringBuilder($"Message: {Message}\nStatusCode: {StatusCode}, ErrorCode: {ErrorCode}, IsTransient: {IsTransient}");
```

File: e2e/LongHaul/service/DeviceOperations.cs [hunk 3]
```diff
@@ -52,27 +54,41 @@ public async Task InvokeDirectMethodAsync(CancellationToken ct)
...
+                        break;
                     }
 
-                    _logger.Trace($"Response status: {response.Status}, payload:\n\t{JsonConvert.SerializeObject(response.PayloadAsString)}", TraceSeverity.Information);
-                }
-                catch (IotHubServiceException ex) when (ex.ErrorCode == IotHubServiceErrorCode.DeviceNotOnline)
-                {
-                    _logger.Trace($"Caught exception invoking direct method {ex}", TraceSeverity.Warning);
+                    // retry delay
+                    await Task.Delay(s_retryInterval, ct).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (drwill-ms) [reply e2e/LongHaul/service/DeviceOperations.cs:88]: This is the only one we had an explicit catch for, so I added a retry for it. The others simply loop on any unrecognized exception.

What I'm looking for is for us to rely on client retry, and implement a retry override for conditions we run into that the SDK doesn't consider retryable.

**Review by tmahmood-microsoft — APPROVED** [review]

**Pull request merged**